index.js

```diff
@@ -21,132 +142,82 @@ var TruffleSchema = {
+        sources = sources.concat(property.additionalSources);
       }
 
-      try {
-        // Will throw an error if key == address and address doesn't exist.
-        value = extra_options[key];
-
-        if (value != undefined) {
-          normalized[key] = value;
-        }
-      } catch (e) {
-        // Do nothing.
+      // iterate over sources until value is defined or end of list met
+      for (var i = 0; value === undefined && i < sources.length; i++) {
+        var source = sources[i];
+        value = source(objectable);
       }
-    });
 
-    Object.keys(deprecated_key_mappings).forEach(function(deprecated_key) {
-      var mapped_key = deprecated_key_mappings[deprecated_key];
-
-      if (normalized[mapped_key] == null) {
-        normalized[mapped_key] = options[deprecated_key] || extra_options[deprecated_key];
+      // run source-agnostic transform on value
+      // (e.g. make sure bytecode begins 0x)
+      if (property.transform) {
+        value = property.transform(value);
       }
-    });
 
-    if (typeof normalized.abi == "string") {
-      normalized.abi = JSON.parse(normalized.abi);
-    }
+      // add resulting (possibly undefined) to normalized obj
+      normalized[key] = value;
+    });
 
-    this.copyCustomOptions(options, normalized);
+    // copy custom options
+    this.copyCustomOptions(objectable, normalized);
 
     return normalized;
   },
 
   // Generate a proper binary from normalized options, and optionally
   // merge it with an existing binary.
-  // TODO: This function needs to be renamed. Binary is a misnomer.
-  generateObject: function(options, existing_object, extra_options) {
-    existing_object = existing_object || {};
-    extra_options = extra_options || {};
+  generateObject: function(objectable, existingObjectable, options) {
+    objectable = objectable || {};
+    existingObjectable = existingObjectable || {};
 
-    options.networks = options.networks || {};
-    existing_object.networks = existing_object.networks || {};
+    options = options || {};
 
-    // Merge networks before overwriting
-    Object.keys(existing_object.networks).forEach(function(network_id) {
-      options.networks[network_id] = existing_object.networks[network_id];
-    });
+    obj = this.normalize(objectable);
+    existingObj = this.normalize(existingObjectable);
 
-    var obj = this.normalizeInput(existing_object, options);
+    Object.keys(existingObj).forEach(function(key) {
+      // networks will be skipped because normalize replaces undefined with {}
+      if (obj[key] === undefined) {
+        obj[key] = existingObj[key];
+      }
+    });
 
-    if (options.overwrite == true) {
-      existing_object = {};
-    }
+    Object.keys(existingObj.networks).forEach(function(network_id) {
+      obj.networks[network_id] = existingObj.networks[network_id];
+    });
 
-    obj.contract_name = obj.contract_name || "Contract";
+    // if (options.overwrite == true) {
+    //   e = {};
+    // }
```

Comment:
I'm not entirely sure why the original line is here. I'd take it out (like you did) and obliterate the comment. There might be repercussions but hopefully we can find them. My guess is it's cruft from an old version of the artifactor.